File: 050 Pow(x, n).c
/*
Pow(x,n)
https://leetcode.com/problems/powx-n/

Implement pow(x, n), which calculates x raised to the power n (xn).

Example 1:

Input: 2.00000, 10
Output: 1024.00000
Example 2:

Input: 2.10000, 3
Output: 9.26100
Example 3:

Input: 2.00000, -2
Output: 0.25000
Explanation: 2-2 = 1/22 = 1/4 = 0.25
*/
double myPow(double x, int n)
{
    double answer = 1;
    double base = x;
    long long exponent = n;
    
    if (exponent < 0)
    {
        exponent = -exponent;
        base = 1 / base;
    }
        
    while (exponent != 0)
    {
        /*
        Let i be the iteration count, then
        base is the product component needed on exponent[i]
        */
        if (exponent & 1)
        {
            answer *= base;
        }
        
        exponent >>= 1;
        base *= base;
    }
    
    return answer;
}

File: 050 Pow(x, n).py
'''
50. Pow(x, n)
https://leetcode.com/problems/powx-n/

Implement pow(x, n), which calculates x raised to the power n (i.e., xn).

Example 1:

Input: x = 2.00000, n = 10
Output: 1024.00000

Example 2:

Input: x = 2.10000, n = 3
Output: 9.26100

Example 3:

Input: x = 2.00000, n = -2
Output: 0.25000
Explanation: 2-2 = 1/22 = 1/4 = 0.25

Constraints:
    -100.0 < x < 100.0
    -231 <= n <= 231-1
    n is an integer.
    Either x is not zero or n > 0.
    -104 <= xn <= 104
'''
class Solution:
    def myPow(self, x: float, n: int) -> float:
        result = 1
        base = x if n >=0 else 1 / x
        exp = n if n >= 0 else -n
        # Use binary to achieve constant time
        while exp > 0:
            if exp & 1:
                result *= base
            # The exp shifts right equals square of base
            exp >>= 1
            base *= base
        return result